src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,244 @@
+				}
+			});
+		}
+
+		private void GetEventPosition(string streamName, long eventNumber, IEnvelope envelope) {
+			var streamId = _readIndex.GetStreamId(streamName);
+			var result = _readIndex.ReadEventInfo_KeepDuplicates(streamId, eventNumber);
+
+			var eventPositions = new EventPosition[result.EventInfos.Length];
+
+			for (int i = 0; i < result.EventInfos.Length; i++) {
+				var eventInfo = result.EventInfos[i];
+				var logPos = eventInfo.LogPosition;
+				var chunk = _db.Manager.GetChunkFor(logPos);
+				var localPosition = chunk.ChunkHeader.GetLocalLogPosition(logPos);
+				var chunkEventOffset = chunk.GetActualRawPosition(localPosition);
+
+				// all the events returned by ReadEventInfo_KeepDuplicates() must exist in the log
+				// since the log record was read from the chunk to check for hash collisions.
+				if (chunkEventOffset < 0)
+					throw new Exception($"Failed to fetch actual raw position for event at log position: {logPos}");
+
+				if (chunkEventOffset > uint.MaxValue)
+					throw new Exception($"Actual raw position for event at log position: {logPos} is larger than uint.MaxValue: {chunkEventOffset}");
+
+				eventPositions[i] = new EventPosition(
+					logPosition: logPos,
+					chunkFile: Path.GetFileName(chunk.FileName),
+					chunkVersion: chunk.ChunkHeader.Version,
+					chunkComplete: chunk.ChunkFooter is { IsCompleted: true },
+					chunkEventOffset: (uint) chunkEventOffset);
+			}
+
+			envelope.ReplyWith(
+				new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.Success, eventPositions));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkLock message) {
+			if (_switchChunksLock.TryAcquire())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkUnlock message) {
+			if (_switchChunksLock.TryRelease())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunk message) {
+			ThreadPool.QueueUserWorkItem(_ => {
+				try {
+					SwitchChunk(message.TargetChunkFile, message.NewChunkFile, message.Envelope);
+				} catch (Exception ex) {
+					Log.Error(ex, "An error has occurred when trying to switch chunk: {targetChunk} with chunk: {newChunk}.",
+						message.TargetChunkFile, message.NewChunkFile);
+					message.Envelope.ReplyWith(
+						new RedactionMessage.SwitchChunkCompleted(SwitchChunkResult.UnexpectedError));
+				}
+			});
+		}
+
+		private void SwitchChunk(string targetChunkFile, string newChunkFile, IEnvelope envelope) {
```

Comment:
it's just slightly tricky here to be sure that we dont reply twice (i wonder if we have (or should have) some coding standard around whether envelopes should deal with multiple replies or whether the person replying should be responsible for not replying more than once (unless they are supposed to). the actual implementation of the envelope calls Set rather than TrySet on the tcs so it will be sad if we do reply more than once currently

Reply:
hmm yup true, it would help make the code easier to read

for this case, an easy way to reason about it is that the envelope reply is the last line of code that's executed before the function returns from the thread created in `Handle(RedactionMessage.SwitchChunk message)`